### PR TITLE
Feature/oauth

### PR DESCRIPTION
**Pull request by AllKinds**
Hard link to https://github.com/meateam/file-service/pull/203

### CODE REVIEW

**Comment by Shahar-Y** [comment]
Also, please approve that it was tested in the kubernetes

File: oauth/oauth.go [hunk 1]
```diff
@@ -74,130 +107,250 @@ func NewOAuthMiddleware(
...
+
+// validateRequiredScope checks if there is a specific scope in the context (unless it is the drive client).
+func (m *Middleware) ValidateRequiredScope(ctx *gin.Context, requiredScope string) bool {
+
+	appID := ctx.Value(ContextAppKey)
+	if appID == DriveAppID {
+		return true
+	}
+
+	// temporary
```
Comment (Shahar-Y) [inline oauth/oauth.go:332]: ?

File: oauth/oauth.go [hunk 2]
```diff
@@ -74,130 +107,250 @@ func NewOAuthMiddleware(
...
+	authType := ctx.Value(ContextAuthType)
+	if authType == DropboxAuthTypeValue {
+		return true
+	}
+	// temporary
```
Comment (Shahar-Y) [inline oauth/oauth.go:337]: ?

File: oauth/oauth.go (hunk 1 above)
Comment (Shahar-Y) [inline oauth/oauth.go:335]:
```suggestion
	//  temporary
```

File: oauth/oauth.go [hunk 3]
```diff
@@ -38,6 +38,42 @@ const (
 
 	// ServiceAuthTypeValue is the value of service for AuthTypeHeader key
 	ServiceAuthTypeValue = "Service"
+
+	// ServiceAuthTypeValue is the value of service for AuthTypeHeader key
```
Comment (Shahar-Y) [inline oauth/oauth.go:42]: Bad comment

File: oauth/oauth.go [hunk 4]
```diff
@@ -38,6 +38,42 @@ const (
 
 	// ServiceAuthTypeValue is the value of service for AuthTypeHeader key
 	ServiceAuthTypeValue = "Service"
```
Comment (Shahar-Y) [inline oauth/oauth.go:40]: Add a comment which says that this will be deprecated

Comment (Shahar-Y) [reply oauth/oauth.go:40]: Also, remove `OutAdminScope`. It's irrelevant.

File: oauth/oauth.go [hunk 5]
```diff
@@ -69,135 +105,256 @@ func NewOAuthMiddleware(
...
-			return
+func (m *Middleware) AuthorizationScopeMiddleware(requiredScope string) gin.HandlerFunc {
+	return func(ctx *gin.Context) {
+		authType := ctx.GetHeader(AuthTypeHeader)
+		ctx.Set(ContextAuthType, authType)
+
+		var err error
+
+		switch authType {
+		case ServiceAuthTypeValue:
```
Comment (Shahar-Y) [inline oauth/oauth.go:121]: Add comment that this header value is deprecated

**Review by Shahar-Y — CHANGES_REQUESTED** [review]

File: oauth/oauth.go (hunk 3 above)
Comment (AllKinds) [inline oauth/oauth.go:42]:
```suggestion
	// DropboxAuthTypeValue is the value of the AuthTypeHeader key for the Dropbox services
```

File: oauth/oauth.go (hunk 5 above)
Comment (AllKinds) [inline oauth/oauth.go:121]:
```suggestion
              // DEPRECATED: this case should be delete in the next version 
		case ServiceAuthTypeValue:
```

File: oauth/oauth.go [hunk 6]
```diff
@@ -69,135 +105,256 @@ func NewOAuthMiddleware(
...
+	if appID == DriveAppID {
+		return true
+	}
+
+	// temporary dropbox
+	authType := ctx.Value(ContextAuthType)
+	if authType == DropboxAuthTypeValue {
+		return true
+	}
+	// temporary
```
Comment (AllKinds) [inline oauth/oauth.go:341]: This code's purpose is to allow Dropbox to preform any action it wants (in the name of an external user) without validating that it (Dropbox) has the corresponding scopes.
This should be deleted after successfully integrating with Dropbox (with an emphasis on the correct scopes).

File: oauth/oauth.go (hunk 2 above)
Comment (AllKinds) [reply oauth/oauth.go:341]: replied in the next comment

File: oauth/oauth.go [hunk 7]
```diff
@@ -69,135 +105,257 @@ func NewOAuthMiddleware(
...
-			return
+func (m *Middleware) AuthorizationScopeMiddleware(requiredScope string) gin.HandlerFunc {
+	return func(ctx *gin.Context) {
+		authType := ctx.GetHeader(AuthTypeHeader)
+		ctx.Set(ContextAuthType, authType)
+
+		var err error
+
+		switch authType {
+              // DEPRECATED: this case should be delete in the next version
```
Comment (AllKinds) [inline oauth/oauth.go:121]:
```suggestion
              // DEPRECATED: this case should be deleted in the next version 
```

File: server/auth/auth.go [hunk 8]
```diff
@@ -43,6 +44,9 @@ const (
 	// ServiceAuthTypeValue is the value of service for AuthTypeHeader key
```
Comment (AllKinds) [inline server/auth/auth.go:44]:
```suggestion
	// DEPRECATED: ServiceAuthTypeValue is the value of service for AuthTypeHeader key
```

File: oauth/oauth.go [hunk 9]
```diff
@@ -38,6 +38,42 @@ const (
 
 	// ServiceAuthTypeValue is the value of service for AuthTypeHeader key
```
Comment (AllKinds) [inline oauth/oauth.go:39]:
```suggestion
	// DEPRECATED: ServiceAuthTypeValue is the value of service for AuthTypeHeader key
```

File: oauth/oauth.go (hunk 5 above)
Comment (AllKinds) [reply oauth/oauth.go:118]: added

File: oauth/oauth.go (hunk 4 above)
Comment (AllKinds) [reply oauth/oauth.go:36]: done

File: oauth/oauth.go (hunk 3 above)
Comment (AllKinds) [reply oauth/oauth.go]: edited

**Pull request merged**